beacon-chain/das/availability.go

```diff
@@ -0,0 +1,250 @@
+	if len(missing) == 0 {
+		return nil
+	}
+	// Some commitments are missing, and we know they aren't in the cache because there are no matching commitments in the cache.
+	// This means DA check fails (at this time).
+	if m == 0 {
+		return NewMissingIndicesError(missing)
+	}
+
+	// If we have any of the missing commitments in the cache, collect them to be verified and persisted.
+	bisect := make([]*ethpb.BlobSidecar, 0, len(missing))
+	for i := range missing {
+		idx := missing[i]
+		if cachedScs[idx] == nil {
+			continue
+		}
+		bisect = append(bisect, cachedScs[idx])
+	}
+
+	// persistIfProven returns the updated value of dbidx, which we can use to check if any comitments are still missing.
+	dbidx, err = s.persistIfProven(ctx, entry, bisect)
+	if err != nil {
+		return err
+	}
+
+	missing, err = dbidx.missing(blockCommitments)
+	if err != nil {
+		return err
+	}
+	// If there are still missing commitments, DA check fails (for now).
+	if len(missing) > 0 {
+		return NewMissingIndicesError(missing)
+	}
+
+	// no more missing - success!
+	return nil
+}
+
+// persistIfProven saves all of the given BlobSidecars that pass proof verification. The BlobSidecars own stated commitment is used
+// as the commitment to check against, so it is critical this method only be used where the KzgCommitment for a BlobSidecar has been
+// matched against the commitment at the corresponding index in the block's set of commitments.
+func (s *CachingDBVerifiedStore) persistIfProven(ctx context.Context, entry *cacheEntry, scs []*ethpb.BlobSidecar) (dbidx, error) {
+	failed := make(map[[48]byte]struct{})
+
+	err := kzg.BisectBlobSidecarKzgProofs(scs)
+	if err != nil {
+		var errFailed *kzg.KzgProofError
+		// Can't cast to KzgProofError to see which commitments failed
+		if !errors.As(err, errFailed) {
+			return nil, err
+		}
+
+		// The custom error returned by BisectBlobSidecarKzgProofs provides the Failed method to retrieve a list of commitments
+		// that failed verification after bisection. Use this to filter out BlobSidecars and only save the ones that were successfully
+		// proven.
+		for _, fc := range errFailed.Failed() {
+			failed[fc] = struct{}{}
+		}
+	}
+
+	save := make([]*ethpb.BlobSidecar, 0, len(scs)-len(failed))
+	for i := range scs {
+		car := bytesutil.ToBytes48(scs[i].KzgCommitment)
+		// Skipping failed commitments.
+		if _, ok := failed[car]; ok {
+			log.WithField("block_root", fmt.Sprintf("%#x", scs[i].BlockRoot)).WithField("index", scs[i].Index).
+				WithField("commitment", fmt.Sprintf("%#x", scs[i])).WithError(err).
+				Error("commitment proof failure. evicting BlobSidecar from cache and not saving to db")
+			continue
+		}
+		// All other BlobSidecars were successfully proven, so they should be saved to the db.
+		save = append(save, scs[i])
+	}
+	// Save any of the Sidecars where the commitments passed proof verification.
+	if err := s.db.SaveBlobSidecar(ctx, save); err != nil {
```

Comment:
IIUC here, we save selective and verified blob sidecar to DB even if some commitments are missing. Is this what we want? I thought what we want is more all-or-nothing